Set of essential and/or useful OBS Studio extensions

kakra opened the pull request:
This is a set of OBS extensions I'm personally using and are probably useful or essential to any Gentoo user running OBS Studio.

All extensions have been runtime checked with X11 (used in production), and build-time checked with Wayland.

Most dependencies are requiring a minimum version simply because that's what I tested with.

v2: Corrected the package licenses

v3: Work around some plugins installing the solib twice in two different locations

<!-- Please put the pull request description above -->

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

There's  the following message which is probably okay:
```
media-plugins/obs-vkcapture
  UnstableOnly: for arches: [ amd64, arm64 ], all versions are unstable: [ 1.5.1 ]
  RequiredUseDefaults: version 1.5.1: profile: 'default/linux/amd64/23.0' (38 total) failed REQUIRED_USE: ( X || wayland )
```

kostadinsh commented:
You are missing multilib support for obs-vkcapture, see how it's done in the package in guru: https://github.com/gentoo/guru/tree/master/media-video/obs-vkcapture

kakra commented:
> You are missing multilib support for obs-vkcapture, see how it's done in the package in guru: https://github.com/gentoo/guru/tree/master/media-video/obs-vkcapture

OBS does not provide multilib builds. The ebuild fails with not having a 32bit libobs, if I try that.

immolo commented:
Any chance of adding `media-video/obs-multi-rtmp` in with a bump into this PR please? After that it will get my vote and a needed addition to Gentoo.

kakra commented:
@immolo I have no use for that plugin but if you're willing to test, I can try adding it.

immolo commented:
> @immolo I have no use for that plugin but if you're willing to test, I can try adding it.

Understood, I'll pick this up in separate PR and I'll take ownership of it from there.

kakra commented:
> Understood, I'll pick this up in separate PR and I'll take ownership of it from there.

I think OBS plugins should go to "media-plugins" and not "media-video", btw. What do you think?

immolo commented:
> > Understood, I'll pick this up in separate PR and I'll take ownership of it from there.
> 
> I think OBS plugins should go to "media-plugins" and not "media-video", btw. What do you think?

I was thinking the same actually, only keeping it as the status quo for histrionic reasons. The fact we both think the same is enough to convince me to do the same.

kakra commented:
This PR was lurking long enough here that we could join forces and merge our two PRs here. If you send a PR to my branch, I can merge it to the list of commits. This should increase the chance of both becoming merged. OTOH, I'm thinking of moving it to GURU or keep a copy of it in a separate overlay. I'll keep the plugins updated every now and then.

I considered using obs-multi-rtmp because I do actually use multiple outputs, but that plugin didn't actually do what I needed: I wanted to record a stream with some of the overlays excluded. And that is what obs-source-record does for me: it can record any scene or source individually. I think obs-source-record can actually stream to multiple outputs too, not just record. But it may be a little more inconvenient for that.

Chiitoo commented:
I do at least from time to time remember that this exists, and that I want to look into it, but I do admit I kind of forgot if we decided anything with regards to GURU, heh.

In any case, I'll see what I can do hopefully sooner, rather than later, but if anyone else is interested in these too, don't be waiting for me.  :]

immolo commented:
> I think obs-source-record can actually stream to multiple outputs too, not just record. But it may be a little more inconvenient for that.

That does seem useful, I'm not entirely sure it does what I need though. Not sure GH is the place to discuss this  but I'm happy to talk more on IRC if you feel a need.

I don't want to put words into a developer's mouth, so I feel now we should just watch this space for both PRs, see #39508 

I will add though that I have seen users asking for the plugins you are willing to maintain.

kakra commented:
> For `media-plugins/obs-pipewire-audio-capture-1.1.5` I get

This is surprising... I've just tried myself, and while it worked previously, I got the same error using this command:
```bash
ebuild obs-pipewire-audio-capture-1.1.5.ebuild clean install
```

But I think I messed up somewhere and copied an ebuild from another package without properly adjusting it. Will fix.

> And for media-plugins/obs-source-dock-0.4.1

It does work here with obs-studio-31.0.0-r1. Maybe it does not support older versions? Can you confirm this?

Thanks for taking your time with this. :-)

kakra commented:
> It does work here with obs-studio-31.0.0-r3. Maybe it does not support older versions? Can you confirm this?

I've raised the version requirement. I'm pretty sure I did not test this with OBS below v31.

Chiitoo commented:
A-ha!

It seems to build with `USE="wayland"` but fails with `USE="-wayland"`.

Chiitoo commented:
Also, `qa-vdb` picked up a couple of things to maybe check a look at:

```
VDB: detected possibly incorrect RDEPEND (media-plugins/obs-source-dock-0.4.1)
 > dev-qt/qtbase:6

VDB: detected possibly incorrect RDEPEND (media-plugins/obs-vkcapture-1.5.1)
                         > dev-libs/wayland
media-libs/vulkan-loader <
x11-libs/libX11          <
x11-libs/libxcb          | x11-libs/libxcb:=
```

kakra commented:
> What might the commented out patch be?

There's been a problem in the code < 0.4.1 which fails to compile in Gentoo due to warnings. The author since fixed it and I dropped my patch.

I'll check the qa-vdb warnings. How did you get those?

kakra commented:
I've tried:
```bash
for i in media-plugins/obs-*/*.ebuild; do sudo USE="X wayland" ebuild $i clean install || exit; done
for i in media-plugins/obs-*/*.ebuild; do sudo USE="-X" ebuild $i clean install || exit; done
for i in media-plugins/obs-*/*.ebuild; do sudo USE="-wayland" ebuild $i clean install || exit; done
```
(using sudo because of egit permission issues)

All went well but I could not reproduce your issue with `-wayland` in the first place. But the build system is designed in a way that detects wayland and X during configuration, and I cannot easily work around that. Thus, we need to ensure that all deps are installed before building. So maybe still missing deps?

I also reviewed the direct deps of the installed solibs and updated the deps in the build (obs-vkcapture).

Finally, I fixed the year 2025 warnings.

See https://github.com/gentoo/gentoo/compare/0b94b01fe30df79de72fef465b7cdfea78c4d5ed..99b3631688e4101ed77542cb42b60616a02d7500

If there are QA warnings let me know how or where to find them.

Chiitoo commented:
Yeah, I noticed the upstream issue and PR, and sorry I wasn't very specific about the `USE="wayland"`.

What I meant was that `media-video/obs-studio` itself needs to be built without `wayland` support, and I suppose this all happens because `obs-nix-platform.h` is only included if `ENABLE_WAYLAND` is defined [1].

Version `30.2.3-r1` also works at least build-wise, but it will probably not be around for long in any case.

1. https://github.com/exeldro/obs-source-dock/blob/master/qt-display.cpp#L16

kakra commented:
> What I meant was that `media-video/obs-studio` itself needs to be built without `wayland` support, and I suppose this all happens because `obs-nix-platform.h` is only included if `ENABLE_WAYLAND` is defined

Then I don't see how we could easily resolve this. The package in question looks for wayland packages in the system and assumes that OBS is built with wayland support.

I can only file a bug report about that because I cannot force the build script to build without wayland.

How should we proceed? Do you want to cherry pick all the packages which are fine, and I rebase? Or should I create a follow-up PR with the problematic packages excluded, then rebase this one?

Chiitoo commented:
I suppose could just depend on 'obs-studio[wayland]' for that package, if it does not support non-wayland set-ups.

kakra commented:
> But the build system is designed in a way that detects wayland and X during configuration, and I cannot easily work around that. Thus, we need to ensure that all deps are installed before building. So maybe still missing deps?

Actually I confused this with obs-vkcapture which - from CMakeLists.txt - should have a similar issue. But you didn't get any issue there, did you?

The build system of obs-source-dock doesn't care about wayland at all. I think for the moment, we can just do as you suggest and depend hard on OBS with wayland support...

Chiitoo commented:
> Actually I confused this with obs-vkcapture which - from CMakeLists.txt - should have a similar issue. But you didn't get any issue there, did you?

Indeed, `obs-vkcapture` seems to be fine with `obs-studio[-wayland]`.

Chiitoo commented:
Sorry, forgot about `qa-vdb`.

It's from `app-portage/iwdevtools`, and comes with the following output for the latest:

```
 VDB: detected possibly incorrect RDEPEND (media-plugins/obs-pipewire-audio-capture-1.1.5)
media-video/wireplumber <
                        > media-video/pipewire:=
VDB: detected possibly incorrect RDEPEND (media-plugins/obs-vkcapture-1.5.1)
media-libs/vulkan-loader <
x11-libs/libX11          <
x11-libs/libxcb          | x11-libs/libxcb:=
```

The suggestions are not 100% correct always, depending on the packages and stuff, but it is quite handy.

I also use the following Gentoo flag for LD:

```
LDFLAGS="${LDFLAGS} -Wl,--defsym=__gentoo_check_ldflags__=0"
```

kakra commented:
Thanks, yes, `va-qdb` is really useful. As far as I see it makes decoupling DEPEND and RDEPEND much more easy. I've adjusted the ebuilds and verified with `lddtree`, looks good now.

I'm not sure what your `LDFLAGS` do, tho. As far as I can see, it's just about disabling checks for QA to find packages that ignore Gentoo `LDFLAGS`? So for this, I probably don't want to set this as otherwise QA won't warn about ignored `LDFLAGS`?

Chiitoo commented:
Ie believe that flag should tell us if a package is not respecting `LDFLAGS`, but right now I can't quite seem to find it documented anywhere.

I also can't find the IRC conversation from where I got it... right now...
